### PR TITLE
Fix VarAnnotationIncorrectNullableRector, add support for closure params when generating php doc

### PR DESCRIPTION
**Pull request by dorrogeray**
Hi,

this should be the fix for https://github.com/rectorphp/rector-src/pull/2920

I have done two things:

- If there is a closure which should be nullable, the null will be prepended at the beginning of the phpdoc union type, to improve readability
- I have added support for closure parameters to be generated for phpstan closure type when generating phpdoc

Sorry about the cognitive complexity breach, but I don't see a simple way to simplify that function without forcefully splitting it into smaller methods, which would not improve the readability imho in this particular case, but feel free to suggest some refactoring.

### CODE REVIEW

File: packages/PHPStanStaticTypeMapper/TypeMapper/ClosureTypeMapper.php [hunk 1]
```diff
@@ -43,7 +43,15 @@ public function mapToPHPStanPhpDocTypeNode(Type $type, string $typeKind): TypeNo
...
-        return new SpacingAwareCallableTypeNode($identifierTypeNode, [], $returnDocTypeNode);
+        $parameterDocTypeNodes = [];
+        foreach ($type->getParameters() as $parameter) {
+            $parameterDocTypeNodes[] = $this->phpStanStaticTypeMapper->mapToPHPStanPhpDocTypeNode(
+                $parameter->getType(),
+                $typeKind
+            );
+        }
+
+        return new SpacingAwareCallableTypeNode($identifierTypeNode, $parameterDocTypeNodes, $returnDocTypeNode);
```
Comment (samsonasik) [inline packages/PHPStanStaticTypeMapper/TypeMapper/ClosureTypeMapper.php:54]: could you move it to private method and call it in return? that's should solve cognitive complexity phpstan notice

File: rules/TypeDeclaration/Helper/PhpDocNullableTypeHelper.php [hunk 2]
```diff
@@ -133,29 +134,38 @@ private function resolveUpdatedPhpDocTypeFromPhpDocTypeAndPhpParserTypeNullInfo(
     ): ?Type {
         /** @var array<(NullType | UnionType)> $updatedDocTypes */
         $updatedDocTypes = [];
-        $phpDocTypeContainsNullType = false;
+        $isPhpDocTypeContainingNullType = false;
+        $isPhpDocTypeContainingClosureType = $phpDocType instanceof ClosureType;
         if ($phpDocType instanceof UnionType) {
-            $phpDocTypeContainsNullType = TypeCombinator::containsNull($phpDocType);
+            $isPhpDocTypeContainingNullType = TypeCombinator::containsNull($phpDocType);
             foreach ($phpDocType->getTypes() as $subType) {
```
Comment (samsonasik) [inline rules/TypeDeclaration/Helper/PhpDocNullableTypeHelper.php]: You can create to separate 2 private methods:

- `hasClosureType()` which loop break on `$subType instanceof ClosureType` return early
- `resolveUpdatedDocType` which loop and append existing `$updatedDocTypes[]` when `$subType` is not `NullType`

so variable can be assigned.

File: rules/TypeDeclaration/Helper/PhpDocNullableTypeHelper.php [hunk 3]
```diff
@@ -133,29 +134,38 @@ private function resolveUpdatedPhpDocTypeFromPhpDocTypeAndPhpParserTypeNullInfo(
...
             return null;
         }
 
         if ($isPhpParserTypeContainingNullType) {
-            $updatedDocTypes[] = new NullType();
+            if ($isPhpDocTypeContainingClosureType) {
+                array_unshift($updatedDocTypes, new NullType());
+            } else {
+                $updatedDocTypes[] = new NullType();
+            }
```
Comment (samsonasik) [inline rules/TypeDeclaration/Helper/PhpDocNullableTypeHelper.php]: you can create separate private method for this, return updated `$pudatedDocTypes`

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you 😊

**Pull request merged**